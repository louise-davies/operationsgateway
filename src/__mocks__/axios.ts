// TODO: move __mocks__ folder back to package root once facebook/create-react-app#7539 is fixed

const requests = {
  get: jest.fn((path) => {
    if (path === '/operationsgateway-settings.json') {
      return Promise.resolve({
        data: {
          apiUrl: 'api',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
};

export default requests;
