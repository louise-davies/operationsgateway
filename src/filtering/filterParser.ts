type TokenType =
  | 'not'
  | 'and'
  | 'or'
  | 'number'
  | 'string'
  | 'compop'
  | 'openparen'
  | 'closeparen'
  | 'channel'
  | 'unaryop';

export interface Token {
  type: TokenType;
  value: string;
  displayValue?: string;
}

export const operators: Token[] = [
  {
    type: 'not',
    value: 'not',
  },
  {
    type: 'and',
    value: 'and',
  },
  {
    type: 'or',
    value: 'or',
  },
  {
    type: 'openparen',
    value: '(',
  },
  {
    type: 'closeparen',
    value: ')',
  },
  {
    type: 'compop',
    value: '=',
  },
  {
    type: 'compop',
    value: '!=',
  },
  {
    type: 'compop',
    value: '<=',
  },
  {
    type: 'compop',
    value: '>=',
  },
  {
    type: 'compop',
    value: '<',
  },
  {
    type: 'compop',
    value: '>',
  },
  {
    type: 'unaryop',
    value: 'is not null',
  },
  {
    type: 'unaryop',
    value: 'is null',
  },
];

/**
 * Converts an operator token into a MongoDB operator
 * @param opToken Operator token to convert
 * @returns string with the corresponding MongoDB operator
 */
const convertOperator = (opToken: Token): string => {
  if (opToken.type !== 'compop' && opToken.type !== 'unaryop') {
    throw new ParserError(`Error converting operator ${opToken.value}`);
  } else {
    if (opToken.value === '=' || opToken.value === 'is null') {
      return 'eq';
    }
    if (opToken.value === '!=' || opToken.value === 'is not null') {
      return 'ne';
    }
    if (opToken.value === '<') {
      return 'lt';
    }
    if (opToken.value === '>') {
      return 'gt';
    }
    if (opToken.value === '<=') {
      return 'lte';
    }
    if (opToken.value === '>=') {
      return 'gte';
    } else {
      throw new ParserError(`Error converting operator ${opToken.value}`);
    }
  }
};

/**
 * Converts the short name of a channel into the DB name
 * @param channel channel to convert
 * @returns Database channel name
 * @example
 * // returns metadata.timestamp
 * convertChannel("timestamp")
 * @example
 * // returns channels.CHANNEL_1.data
 * convertChannel("CHANNEL_1")
 */
const convertChannel = (channel: string): string => {
  if (
    ['timestamp', 'shotnum', 'activeArea', 'activeExperiment'].includes(channel)
  ) {
    return `metadata.${channel}`;
  } else {
    return `channels.${channel}.data`;
  }
};

export class ParserError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ParserError';
  }
}

/** Class representing a stream of tokens, which keeps track of where in the array we've processed up to */
class Input {
  tokens: Token[];
  pos: number;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
    this.pos = 0;
  }

  /**
   * Returns the token at the current position relative to the given offset
   * @param off the offset to look at
   * @returns a Token if it's within the range of the stream otherwise null
   */
  public peek(off: number): Token | null {
    const n = this.pos + off;
    if (n >= 0 && n < this.tokens.length) {
      return this.tokens[this.pos + off];
    } else {
      return null;
    }
  }

  /**
   * Returns the next token in the stream and moves the position forward.
   * @param types an array of `TokenTypes` check the current token against
   * @returns The next token in the stream or null if it's the end of the stream
   * @throws {ParserError} if the token does not match the `types` param or if we attempt to read beyond the end of the stream
   */
  public consume(types?: TokenType[]): Token | null {
    if (types) {
      if (this.pos < this.tokens.length) {
        if (types.includes(this.tokens[this.pos].type)) {
          return this.tokens[this.pos++];
        } else {
          throw new ParserError(
            `Expected token from types [${types}] at token ${
              this.peek(0)?.value
            }` // TODO: better error message?
          );
        }
      } else {
        throw new ParserError(`Missing token at end. Expected: ${types}`);
      }
    } else {
      if (this.pos < this.tokens.length) {
        return this.tokens[this.pos++];
      } else if (this.pos === this.tokens.length) {
        return null;
      } else {
        throw new ParserError('Attempt to read beyond end');
      }
    }
  }
}

/**
 * Class representing a predicate e.g. `CHANNEL_1 < 1` or `activeExperiment is not null`
 *
 * Predicate ::= channel COMPOP param | channel UNARYOP
 */
class Predicate {
  param1: string | number | undefined;
  param2: string | number | undefined;
  op: Token | undefined;

  /**
   * @param input `Input` stream to use
   * @throws {ParserError} if the predicate has invalid syntax e.g. number on LHS, no operator etc.
   */
  constructor(input: Input) {
    let token = input.peek(0);
    if (token !== null) {
      if (token.type === 'channel') {
        this.param1 = convertChannel(token.value);
        input.consume();
      } else if (token.type === 'string') {
        throw new ParserError(
          `Unexpected string on left hand side of expression: ${token.value}`
        );
      } else if (token.type === 'number') {
        throw new ParserError(
          `Unexpected number on left hand side of expression: ${token.value}`
        );
      } else {
        throw new ParserError(`Unexpected: ${token.value}`);
      }
    } else {
      throw new ParserError('Missing operand');
    }
    this.op = input.consume(['compop', 'unaryop']) as Token;
    if (this.op.type === 'compop') {
      token = input.peek(0);
      if (token !== null) {
        if (token.type === 'channel') {
          this.param2 = convertChannel(token.value);
          input.consume();
        } else if (token.type === 'string') {
          // remove quotes
          this.param2 = token.value.replaceAll('"', '').replaceAll("'", '');
          input.consume();
        } else if (token.type === 'number') {
          this.param2 = Number(token.value);
          input.consume();
        } else {
          throw new ParserError(`Unexpected: ${token.value}`);
        }
      } else {
        throw new ParserError('Missing operand');
      }
    }
  }

  public toString(): string {
    let s = '';
    if (this.param1 && this.op && this.op.type === 'compop' && this.param2) {
      const param2 =
        typeof this.param2 === 'string' ? `"${this.param2}"` : this.param2;
      s = `{"${this.param1}":{"$${convertOperator(this.op)}":${param2}}}`;
    } else if (this.param1 && this.op && this.op.type === 'unaryop') {
      s = `{"${this.param1}":{"$${convertOperator(this.op)}":null}}`;
    }
    return s;
  }
}

/**
 * Class representing a potential NOT expression or parenthesis
 *
 * BooleanFactor ::= ("NOT")? Predicate | ( "(" SearchCondition ")" )
 */
class BooleanFactor {
  not = false;
  predicate: Predicate | undefined;
  searchCondition: SearchCondition | undefined;

  /**
   * @param input `Input` stream to use
   * @throws {ParserError} if the expression has invalid syntax
   */
  constructor(input: Input) {
    let token = input.peek(0);
    if (token == null) {
      throw new ParserError('Expression incomplete');
    }
    if (token.type === 'not') {
      input.consume();
      this.not = true;
    }
    token = input.peek(0);
    if (token == null) {
      throw new ParserError('Expression incomplete');
    }
    if (token.type === 'openparen') {
      input.consume();
      this.searchCondition = new SearchCondition(input);
      input.consume(['closeparen']);
    } else {
      this.predicate = new Predicate(input);
    }
  }

  public toString(): string {
    let s = '';
    if (this.not) {
      s += '{"$not":';
    }
    if (this.predicate) {
      s += this.predicate;
    }
    if (this.searchCondition) {
      s += this.searchCondition;
    }
    if (this.not) {
      s += '}';
    }
    return s;
  }
}

/**
 * Class representing a potential AND expression
 *
 * BooleanTerm ::= BooleanFactor ( "AND" BooleanFactor ) *
 */
class BooleanTerm {
  factors: BooleanFactor[] = [];

  /**
   * @param input `Input` stream to use
   * @throws {ParserError} if the expression has invalid syntax
   */
  constructor(input: Input) {
    this.factors.push(new BooleanFactor(input));
    let token = null;
    while ((token = input.peek(0)) !== null) {
      if (token.type === 'and') {
        input.consume();
        this.factors.push(new BooleanFactor(input));
      } else {
        return;
      }
    }
  }

  public toString(): string {
    let s = '';
    if (this.factors.length === 0) {
      // do nothing - should never happen due to error checking
    } else if (this.factors.length === 1) {
      s += this.factors[0];
    } else {
      s += '{"$and":[';
      this.factors.forEach((f) => (s += f + ','));
      s = s.slice(0, -1);
      s += ']}';
    }
    return s;
  }
}

/**
 * Class representing a potential OR expression
 *
 * SearchCondition ::= BooleanTerm ( "OR" BooleanTerm ) *
 */
class SearchCondition {
  booleanTerms: BooleanTerm[] = [];

  /**
   * @param input `Input` stream to use
   * @throws {ParserError} if the expression has invalid syntax
   */
  constructor(input: Input) {
    this.booleanTerms.push(new BooleanTerm(input));
    let token = null;
    while ((token = input.peek(0)) !== null) {
      if (token.type === 'or') {
        input.consume();
        this.booleanTerms.push(new BooleanTerm(input));
      } else {
        return;
      }
    }
  }

  public toString(): string {
    let s = '';
    if (this.booleanTerms.length === 0) {
      // do nothing - should never happen due to error checking
    } else if (this.booleanTerms.length === 1) {
      s += this.booleanTerms[0];
    } else {
      s += '{"$or":[';
      this.booleanTerms.forEach((t) => (s += t + ','));
      s = s.slice(0, -1);
      s += ']}';
    }
    return s;
  }
}

/**
 * Parses a list of tokens into a MongoDB style filter we can pass to the backend
 * @param tokens array of tokens repesenting the filter
 * @returns a MongoDB style filter which can be passed as a `condition` to the backend
 * @throws {ParserError} Will throw an error with a descriptive message if the filter cannot be parsed
 */
export const parseFilter = (tokens: Token[]): string | never => {
  const input = new Input(tokens);
  const searchCondition = new SearchCondition(input);
  return searchCondition.toString();
};
