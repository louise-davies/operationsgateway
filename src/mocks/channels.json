{
  "_id": "123",
  "channels": {
    "CHANNEL_ABCDE": {
      "name": "Channel_ABCDE",
      "path": "/Channels/1/ABCDE",
      "type": "scalar"
    },
    "CHANNEL_BCDEF": {
      "name": "Channel_BCDEF",
      "path": "/Channels/1/BCDEF",
      "type": "image"
    },
    "CHANNEL_CDEFG": {
      "name": "Channel_CDEFG",
      "path": "/Channels/1/CDEFG",
      "type": "waveform",
      "x_units": "X",
      "y_units": "Y"
    },
    "CHANNEL_DEFGH": {
      "name": "Channel_DEFGH",
      "path": "/Channels/2/DEFGH",
      "type": "scalar",
      "notation": "scientific",
      "precision": 1
    },
    "CHANNEL_EFGHI": {
      "name": "Channel_EFGHI",
      "path": "/Channels/2/EFGHI",
      "type": "image"
    },
    "CHANNEL_FGHIJ": {
      "name": "Channel_FGHIJ",
      "path": "/Channels/2/FGHIJ",
      "type": "waveform"
    },
    "CHANNEL_GHIJK": {
      "name": "Channel_GHIJK",
      "path": "/Channels/3/GHIJK",
      "type": "scalar"
    }
  }
}
