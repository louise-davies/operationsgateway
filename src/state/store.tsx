import { configureStore } from '@reduxjs/toolkit';
import OperationsGatewayMiddleware, {
  listenToMessages,
} from './middleware/operationsgateway.middleware';
import configReducer from './slices/configSlice';

export const store = configureStore({
  reducer: {
    config: configReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(OperationsGatewayMiddleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

listenToMessages(store.dispatch);
