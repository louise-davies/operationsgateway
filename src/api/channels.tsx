import React from 'react';
import axios, { AxiosError } from 'axios';
import {
  FullChannelMetadata,
  FullScalarChannelMetadata,
  isChannelMetadataScalar,
  timeChannelName,
} from '../app.types';
import {
  useQuery,
  UseQueryResult,
  UseQueryOptions,
} from '@tanstack/react-query';
import { Column } from 'react-table';
import {
  renderImage,
  roundNumber,
} from '../table/cellRenderers/cellContentRenderers';
import { selectUrls } from '../state/slices/configSlice';
import { useAppSelector } from '../state/hooks';
import { readSciGatewayToken } from '../parseTokens';

interface ChannelsEndpoint {
  channels: {
    [systemName: string]: Omit<FullChannelMetadata, 'systemName'>;
  };
}

// This metadata is always present in every record
export const staticChannels: { [systemName: string]: FullChannelMetadata } = {
  [timeChannelName]: {
    systemName: timeChannelName,
    name: 'Time',
    type: 'scalar',
    path: '/system',
  },
  shotnum: {
    systemName: 'shotnum',
    name: 'Shot Number',
    type: 'scalar',
    path: '/system',
  },
  activeArea: {
    systemName: 'activeArea',
    name: 'Active Area',
    type: 'scalar',
    path: '/system',
  },
  activeExperiment: {
    systemName: 'activeExperiment',
    name: 'Active Experiment',
    type: 'scalar',
    path: '/system',
  },
};

const fetchChannels = (apiUrl: string): Promise<FullChannelMetadata[]> => {
  return axios
    .get<ChannelsEndpoint>(`${apiUrl}/channels`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      const { channels } = response.data;

      if (!channels || Object.keys(channels).length === 0) return [];

      const convertedChannels: FullChannelMetadata[] = Object.entries(
        channels
      ).map(([systemName, channel]) => ({
        systemName,
        ...channel,
      }));

      return [...Object.values(staticChannels), ...convertedChannels];
    });
};

export interface ChannelSummary {
  first_date: string;
  most_recent_date: string;
  recent_sample: { [timestamp: string]: string | number }[];
}

const fetchChannelSummary = (
  apiUrl: string,
  channel: string
): Promise<ChannelSummary> => {
  return axios
    .get(`${apiUrl}/channels/summary/${channel}`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useChannels = <T extends unknown = FullChannelMetadata[]>(
  options?: UseQueryOptions<FullChannelMetadata[], AxiosError, T, string[]>
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery(
    ['channels'],
    (params) => {
      return fetchChannels(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      ...(options ?? {}),
    }
  );
};

export const useChannelSummary = (
  channel: string | undefined
): UseQueryResult<ChannelSummary, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const dataChannel =
    typeof channel !== 'undefined' && !(channel in staticChannels)
      ? channel
      : '';

  return useQuery(
    ['channelSummary', dataChannel],
    (params) => {
      return fetchChannelSummary(apiUrl, dataChannel);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      enabled: dataChannel.length !== 0,
    }
  );
};

export const constructColumns = (channels: FullChannelMetadata[]): Column[] => {
  const myColumns: Column[] = [];

  channels.forEach((channel: FullChannelMetadata) => {
    const newColumn: Column = {
      Header: () => {
        const headerName = channel.name ? channel.name : channel.systemName;
        // Provide an actual header here when we have it
        // TODO: do we need to split on things other than underscore?
        const parts = headerName.split('_');
        const wordWrap = parts.map(
          (part, i) =>
            // \u200B renders a zero-width space character
            // which allows line-break but isn't visible
            part + (i < parts.length - 1 ? '_\u200B' : '')
        );
        return <React.Fragment>{wordWrap.join('')}</React.Fragment>;
      },
      accessor: channel.systemName,
      // TODO: get these from data channel info
      channelInfo: channel,
    };
    if (isChannelMetadataScalar(channel)) {
      newColumn.Cell = ({ value }) =>
        typeof value === 'number' && typeof channel.precision === 'number' ? (
          <React.Fragment>
            {roundNumber(value, channel.precision, channel.notation)}
          </React.Fragment>
        ) : (
          <React.Fragment>{String(value ?? '')}</React.Fragment>
        );
    } else {
      newColumn.Cell = ({ row, value }) =>
        renderImage(
          value,
          `${channel.name ?? channel.systemName} ${
            channel.type
          } for timestamp ${row.values[timeChannelName]}`
        );
    }
    myColumns.push(newColumn);
  });
  return myColumns;
};

export const getScalarChannels = (
  channels: FullChannelMetadata[]
): FullScalarChannelMetadata[] => {
  return channels.filter(
    (channel) => channel.type === 'scalar'
  ) as FullScalarChannelMetadata[];
};

const useScalarChannelsOptions = {
  select: (data: FullChannelMetadata[]) => getScalarChannels(data),
};

const useAvailableColumnsOptions = {
  select: (data: FullChannelMetadata[]) => constructColumns(data),
};

export const useScalarChannels = (): UseQueryResult<
  FullScalarChannelMetadata[],
  AxiosError
> => {
  return useChannels(useScalarChannelsOptions);
};

export const useAvailableColumns = (): UseQueryResult<Column[], AxiosError> => {
  return useChannels(useAvailableColumnsOptions);
};
