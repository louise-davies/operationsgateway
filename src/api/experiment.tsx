import { useQuery, UseQueryResult } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { ExperimentParams } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const fetchExperiment = (apiUrl: string): Promise<ExperimentParams[]> => {
  return axios
    .get(`${apiUrl}/experiments`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useExperiment = (): UseQueryResult<
  ExperimentParams[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['experiments'],

    queryFn: () => fetchExperiment(apiUrl),
  });
};
