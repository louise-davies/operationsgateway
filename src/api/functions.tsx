import {
  UseMutationResult,
  UseQueryResult,
  useMutation,
  useQuery,
} from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import {
  APIFunctionState,
  DataType,
  FunctionOperator,
  ValidateFunctionState,
} from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

export function convertExpressionsToStrings(
  functionStates: ValidateFunctionState[]
): APIFunctionState {
  const channels = functionStates.flatMap((state) => state.channels);
  const uniqueChannels = Array.from(new Set(channels));

  return {
    channels: uniqueChannels,
    functions: functionStates.map((functionState) => ({
      name: functionState.name,
      expression: functionState.expression
        .map((token) => token.value.trim())
        .join(' ')
        // Remove any spaces around the open parenthesis '('
        .replace(/\s*\(\s*/g, '(')
        // Remove any spaces around the close parenthesis ')'
        .replace(/\s*\)\s*/g, ')')
        // Add a space between operators (*, +, -, /) and the following open parenthesis '('
        .replace(/([*+\-/])\(/g, '$1 (')
        // Add a space between the close parenthesis ')' and the following operators (*, +, -, /)
        .replace(/\)([*+\-/])/g, ') $1')
        // Remove any spaces around the double asterisk '**' (exponentiation operator)
        .replace(/\s*\*\*\s*/g, '**'),
    })),
  };
}

const getFunctionsTokens = (apiUrl: string): Promise<FunctionOperator[]> => {
  return axios
    .get(`${apiUrl}/functions/tokens`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useFunctionsTokens = (): UseQueryResult<
  FunctionOperator[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['FunctionTokens'],
    queryFn: () => {
      return getFunctionsTokens(apiUrl);
    },
  });
};

const postValidateFunctions = (
  apiUrl: string,
  functions: ValidateFunctionState[]
): Promise<DataType[]> => {
  const formattedFunctions = convertExpressionsToStrings(functions).functions;

  return axios
    .post(`${apiUrl}/functions/validate`, formattedFunctions, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useValidateFunctions = (): UseMutationResult<
  DataType[],
  AxiosError,
  ValidateFunctionState[]
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useMutation({
    mutationFn: (functions: ValidateFunctionState[]) => {
      return postValidateFunctions(apiUrl, functions);
    },
  });
};
