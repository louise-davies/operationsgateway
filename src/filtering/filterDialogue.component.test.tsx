import React from 'react';
import FilterDialogue from './filterDialogue.component';
import { act, render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  flushPromises,
  getInitialState,
  renderComponentWithProviders,
  testChannels,
  testRecords,
} from '../setupTests';
import { fetchChannels } from '../api/channels';
import { RootState } from '../state/store';
import { PreloadedState } from '@reduxjs/toolkit';
import { Token } from './filterParser';
import axios from 'axios';
import { QueryCache } from 'react-query';

describe('Filter dialogue component', () => {
  let props: React.ComponentProps<typeof FilterDialogue>;
  let promise;

  const createView = (initialState?: PreloadedState<RootState>) => {
    return renderComponentWithProviders(<FilterDialogue {...props} />, {
      preloadedState: initialState,
    });
  };

  beforeEach(() => {
    jest.setTimeout(60000);
    props = {
      open: true,
      onClose: jest.fn(),
    };

    promise = Promise.resolve({ data: testRecords });

    (axios.get as jest.Mock).mockReturnValue(promise);
  });

  it('renders filter dialogue when dialogue is open', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });

  it("doesn't render filter dialogue when dialogue is close", async () => {
    props.open = false;

    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });

  it('calls onClose when close button is clicked', async () => {
    const user = userEvent.setup();
    createView();

    await user.click(screen.getByText('Close'));

    expect(props.onClose).toHaveBeenCalled();
  });

  it('dispatches changeAppliedFilters and onClose when apply button is clicked', async () => {
    const state = {
      ...getInitialState(),
      filter: {
        ...getInitialState().filter,
        appliedFilters: [
          [
            { type: 'channel', value: 'timestamp' },
            { type: 'unaryop', value: 'is not null' },
            { type: 'and', value: 'and' },
            { type: 'channel', value: 'shotnum' },
            { type: 'unaryop', value: 'is null' },
          ],
        ] as Token[][],
      },
    };
    const user = userEvent.setup();

    const { store } = createView(state);

    const filter = screen.getByRole('combobox', { name: 'Filter' });
    await user.type(filter, '{backspace}');
    await user.type(filter, '{backspace}');
    await user.type(filter, '{backspace}');

    expect(screen.getByText('Apply')).not.toBeDisabled();
    await user.click(screen.getByText('Apply'));

    expect(store.getState().filter.appliedFilters).toStrictEqual([
      [
        { type: 'channel', value: 'timestamp' },
        { type: 'unaryop', value: 'is not null' },
      ],
    ]);
    expect(props.onClose).toHaveBeenCalled();
  });

  it('disables apply button when there are errors', async () => {
    const state = {
      ...getInitialState(),
      filter: {
        ...getInitialState().filter,
        appliedFilters: [
          [
            { type: 'channel', value: 'timestamp' },
            { type: 'unaryop', value: 'is not null' },
          ],
        ] as Token[][],
      },
    };
    const user = userEvent.setup();

    createView(state);

    const filter = screen.getByRole('combobox', { name: 'Filter' });
    await user.type(filter, '{backspace}');
    await user.tab();

    expect(screen.getByText('Apply')).toBeDisabled();
  });
});
