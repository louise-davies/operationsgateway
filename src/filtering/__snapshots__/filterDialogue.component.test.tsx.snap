// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter dialogue component doesn't render filter dialogue when dialogue is close 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`Filter dialogue component renders filter dialogue when dialogue is open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthLg MuiDialog-paperFullWidth css-cyxlny-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r0:"
        >
          Filters
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container css-o1xwpy-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-n4yv6w-MuiGrid-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-um3j9i-MuiTypography-root"
              >
                Enter filter
              </h3>
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
              >
                <div
                  class="MuiAutocomplete-root MuiAutocomplete-fullWidth css-12szb0g-MuiAutocomplete-root"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="false"
                      for=":r1:"
                      id=":r1:-label"
                    >
                      Filter
                    </label>
                    <div
                      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiAutocomplete-inputRoot css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-invalid="false"
                        autocapitalize="none"
                        autocomplete="off"
                        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiAutocomplete-input MuiAutocomplete-inputFocused css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                        id=":r1:"
                        role="combobox"
                        spellcheck="false"
                        type="text"
                        value=""
                      />
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-1ftyaf0"
                        >
                          <span>
                            Filter
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-um3j9i-MuiTypography-root"
              >
                Filter help
              </h3>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ejgmvm-MuiTypography-root"
              >
                In the box, start typing data channel names, numbers, mathematical symbols such as 
                &gt;
                 and 
                &lt;=
                 and keywords such as AND and OR. The Wizard will suggest suitable options and indicate using a grey box when each item has been recognised.
              </p>
              <h3
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-um3j9i-MuiTypography-root"
              >
                Operators included
              </h3>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ejgmvm-MuiTypography-root"
              >
                &gt;, &lt;, =, !=, and, or, not
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Apply
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
